ogr/ogrsf_frmts/mssqlspatial/ogrmssqlspatialdatasource.cpp

```diff
@@ -1514,17 +1514,33 @@ OGRSpatialReference *OGRMSSQLSpatialDataSource::FetchSRS(int nId)
     /* -------------------------------------------------------------------- */
     if (poSRS)
     {
-        panSRID = (int *)CPLRealloc(panSRID, sizeof(int) * (nKnownSRID + 1));
-        papoSRS = (OGRSpatialReference **)CPLRealloc(
-            papoSRS, sizeof(void *) * (nKnownSRID + 1));
-        panSRID[nKnownSRID] = nId;
-        papoSRS[nKnownSRID] = poSRS;
-        nKnownSRID++;
+        AddSRIDToCache(nId, poSRS);
     }
 
     return poSRS;
 }
 
+/************************************************************************/
+/*                         AddSRIDToCache()                             */
+/*                                                                      */
+/*      Note: this will not add a reference on the poSRS object. Make   */
+/*      sure it is freshly created, or add a reference yourself if not. */
+/************************************************************************/
+
+void OGRMSSQLSpatialDataSource::AddSRIDToCache(int nId,
+                                               OGRSpatialReference *poSRS)
+{
+    /* -------------------------------------------------------------------- */
+    /*      Add to the cache.                                               */
+    /* -------------------------------------------------------------------- */
+    panSRID = (int *)CPLRealloc(panSRID, sizeof(int) * (nKnownSRID + 1));
+    papoSRS = (OGRSpatialReference **)CPLRealloc(papoSRS, sizeof(void *) *
```

Comment:
@dg0yt What would be the suggested solution? I don't see a performance issue here, especially because we usually look for only a limited set of SRS-s for each data source (a single one in most cases).

Reply:
A ``std::map<int, std::unique_ptr<OGRSpatialReference>> m_oSRSCache`` would be the best C++11 replacement forthe nKnownSRID, panSRID and papoSRS members.

Reply:
Rectification this should be a
``std::map<int, std::unique_ptr<OGRSpatialReference, OGRSpatialReferenceReleaser>> m_oSRSCache``
so that the free function called is OGRSpatialReference::Release() and not plain delete

Reply:
A `std::vector<struct>` with a reasonable `reserve` should be fine, thanks to locality.

Reply:
cf https://github.com/OSGeo/gdal/pull/9119 for a modernization proposal for the PG driver which uses the same pattern